### PR TITLE
(814c) Add caseloads to logged in user

### PR DESCRIPTION
**Pull request by Gweaton**
## Context

https://trello.com/c/PeLVF8dJ/814-ensure-referrers-cant-refer-people-outside-their-active-caseload-ui

We'll shortly be switching over to restricting which people in prison our users can search for, by ensuring they can only search for people in prisons that are part of their caseload. We need to fetch caseloads from the Prison API, which we can then pass as part of our request to the Prisoner Offender Search.

## Changes in this PR

- Tweaks `populateCurrentUser` middleware to only fetch the user once per session, rather than once per page load to cut down on the number of requests to HMPPS Auth (this is an approach taken by a few other services, e.g. CAS 1)
- Fetches the logged-in user's caseloads from the Prison API when populating the current user in the `user` variable. If the user has no caseloads, or the Prison API throws an error, we allow users to continue using the service.
- Stubs the Prison API's caseload endpoint when running the application locally to provide a few standard caseloads for the logged-in user.
- Adds caseload stubs for use in integration tests, as well as a function for setting default stubs.

## Release checklist

[Release process documentation](../doc/how-to/perform-a-release.md)

As part of our continuous deployment strategy we must ensure that this work is
ready to be released once merged.

### Pre-merge

- [ ] There are changes required to the Accredited Programmes API for this change to work...
  - [ ] ... and they have been released to production already

### Post-merge

- [ ] This adds, extends or meaningfully modifies a feature...
  - [ ] ... and I have written or updated an end-to-end test for the happy path in the [Accredited Programmes E2E repo](https://github.com/ministryofjustice/hmpps-accredited-programmes-e2e)
- [ ] This makes new expectations of the API...
  - [ ] ... and I have notified the API developer(s) of changes to the contract tests (Pact), or the API is already compliant
- [ ] [Manually approve](../doc/how-to/perform-a-release.md#releasing-to-the-preprod-environment) release to preprod
- [ ] [Manually approve](../doc/how-to/perform-a-release.md#releasing-to-the-production-environment) release to prod

### CODE REVIEW

File: server/@types/users/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,8 @@
+import type { Caseload } from '@prison-api'
```
Comment (jsrobertson) [inline server/@types/users/index.d.ts:1]: Is it worth adding this file to `paths` in tsconfig so we can import like the others?

**Review by jsrobertson — APPROVED** [review]
Nice one. Just 1 comment about the new users type file.

File: server/@types/users/index.d.ts (hunk 1 above)
Comment (Gweaton) [reply server/@types/users/index.d.ts:1]: Yeah I initially wasn't sure about this as it's only the one class in the file. Thinking about it now though, I _was_ thinking the user types in our codebase are getting a little confusing (when to use each one?) so it'd be good to tidy them up/consolidate them in the near future, which would mean adding more types to that file.

TL;DR, I'll add an alias to pave the way for tidying up users in future.

File: integration_tests/mockApis/caseloads.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+  stubDefaultCaseloads: (): SuperAgentRequest =>
+    stubFor({
+      request: {
+        method: 'GET',
+        url: prisonApiPaths.users.current.caseloads({}),
+      },
+      response: {
+        headers: { 'Content-Type': 'application/json;charset=UTF-8' },
+        jsonBody: defaultCaseloads,
```
[the remark this answers is not in the document]
Comment (Gweaton) [reply integration_tests/mockApis/caseloads.ts:30]: It's imported `as defaultCaseloads` at the top to prevent shadowing the `caseloads` passed into the stub above

File: script/server [hunk 3]
```diff
@@ -31,7 +31,7 @@ fi
 echo "==> Starting the server..."
 
 if [ "$MOCK_API" = true ]; then
-  ACCREDITED_PROGRAMMES_API_URL=http://localhost:9099 PRISONER_OFFENDER_SEARCH_URL=http://localhost:9099 npm run start:dev
+  ACCREDITED_PROGRAMMES_API_URL=http://localhost:9099 PRISONER_OFFENDER_SEARCH_URL=http://localhost:9099 PRISON_API_URL=http://localhost:909 npm run start:dev
```
[the remark this answers is not in the document]
Comment (Gweaton) [reply script/server:34]: Oops, well spotted

File: server/services/userService.ts [hunk 4]
```diff
@@ -1,13 +1,33 @@
-import type { HmppsAuthClient, RestClientBuilderWithoutToken } from '../data'
+import logger from '../../logger'
+import type { UserDetails } from '../@types/users'
```
[the remark this answers is not in the document]
Comment (Gweaton) [reply server/services/userService.ts:2]: ah this was a rebase error, have fixed, thanks!

File: server/middleware/populateCurrentUser.test.ts [hunk 5]
```diff
@@ -83,6 +88,7 @@ describe('populateCurrentUser', () => {
         req = createMock<Request>({
           session: {
             user: {
+              caseloads: [caseloadFactory.build()],
```
[the remark this answers is not in the document]
Comment (Gweaton) [reply server/middleware/populateCurrentUser.test.ts:91]: I could have used `buildList(1)` too, which is a Fishery internal. I don't think it matters all that much in this scenario, will leave this as-is for the time being.

[the remark this answers is not in the document]
Comment (yndajas) [reply server/middleware/populateCurrentUser.test.ts:91]: Yup the between is more if you want to randomise the number to make it more realistically random, but yeah it doesn't really matter here

File: integration_tests/mockApis/caseloads.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (Gweaton) [reply integration_tests/mockApis/caseloads.ts:30]: oh sorry! I see what you mean now. Weirdly it seems to be working for me, but I'll figure out why...

**Review by yndajas — APPROVED** [review]
